### PR TITLE
chore_: create v2 endpoints for status.go to solve EndpointsUnsupported in endpoints.go

### PR DESCRIPTION
**Pull request by qfrank**
I have a dream that the status-mobile can use the status-backend via an HTTP server so that the status-backend can run separately. Now the dream is getting closer. Only the update in status-mobile is left. 

It would be great if you (@ilmotta) could help with the frontend!

relate [status-go PR](https://github.com/status-im/status-go/pull/5847)
relate [comment](https://github.com/status-im/status-go/pull/5847#discussion_r1776928873)

### CODE REVIEW

**Comment by ilmotta** [comment]
> It would be great if you (@ilmotta) could help with the frontend!

Thanks for this PR @qfrank!

-------

@ulisesmac, this is the PR I mentioned to you these days. Would you be interested in making the integration work in the mobile client? Note: this work has less priority than other important bugs affecting the release.

To get started, there's a README in `/cmd/status-backend/README.md`.

```sh
make generate
make run-status-backend PORT=9050
```

@qfrank @igor-sirotin in the README I couldn't find any explanation about how to subscribe to signals. Any tips?

**Comment by qfrank** [comment]
> I couldn't find any explanation about how to subscribe to signals. Any tips?

pls see [this](https://github.com/status-im/status-go/pull/5847/files#diff-afbaf1d50176508a38f662d8cfa4de042ccab10169f8e20ec6f1eaab747779e0) and relate test code @ilmotta

**Comment by ulisesmac** [comment]
> > It would be great if you (@ilmotta) could help with the frontend!
> 
> Thanks for this PR @qfrank!
> 
> @ulisesmac, this is the PR I mentioned to you these days. Would you be interested in making the integration work in the mobile client? Note: this work has less priority than other important bugs affecting the release.
> 
> To get started, there's a README in `/cmd/status-backend/README.md`.
> 
> ```shell
> make generate
> make run-status-backend PORT=9050
> ```
> 
> @qfrank @igor-sirotin in the README I couldn't find any explanation about how to subscribe to signals. Any tips?

I'd love to do it :heart: 
I'll probably be starting with this by Thursday :smile:

File: mobile/status.go
Comment (igor-sirotin) [inline mobile/status.go:288]: Hmm, why unwrap this?

File: protocol/requests/change_database_password.go
Comment (igor-sirotin) [inline protocol/requests/change_database_password.go]: Maybe make it `oldPassword` explicitly?

File: protocol/requests/connection_change.go
Comment (igor-sirotin) [inline protocol/requests/connection_change.go]: I think this can be specific type instead if int?

File: protocol/requests/convert_to_keycard_account.go
Comment (igor-sirotin) [inline protocol/requests/convert_to_keycard_account.go]: Is it possible to use specific type here instead of the json string?

File: protocol/requests/input_connection_string_for_bootstrapping.go
Comment (igor-sirotin) [inline protocol/requests/input_connection_string_for_bootstrapping.go]: Specific type?

**Review by igor-sirotin — COMMENTED** [review]
Nice 👍
I wonder if we can avoid these embedded json strings and use concrete types right away?

**Comment by igor-sirotin** [comment]
> status-mobile can use the status-backend via an HTTP server so that the status-backend can run separately

While this is a good idea for developing, I think for release builds status-go should remain a shared library linked to the app. 

There's no need to use the whole TCP/IP/HTTP stack for this when we can call the method directly. It will be slower with the network and OS involved.

I suggest to make this optional, used only in local debug builds.

cc @ilmotta @qfrank

**Comment by igor-sirotin** [comment]
> I couldn't find any explanation about how to subscribe to signals. Any tips?

Indeed I missed to write this, will do today. In short, you should connect to `ws://localhost:<port>/signals` as the first thing when running status-backend. You'll receive all signals.

cc @ilmotta @ulisesmac

**Comment by ilmotta** [comment]
> There's no need to use the whole TCP/IP/HTTP stack for this when we can call the method directly. It will be slower with the network and OS involved.
> 
> I suggest to make this optional, used only in local debug builds.

Fully agree @igor-sirotin. But, sorry there's always a but :) I wonder (on a very personal level) if running status-go in a separate machine in my LAN would work well.

The overhead could be compensated by some other advantages. I always dreamed of having Status clients running as simpler, thin shells, which could help with some use cases:

- status-go could run much faster and wouldn't be restricted by the hardware the client is running on.
- The overhead of the extra layers could be negligible if latency is low enough (in a LAN for example). Reminds me of Phoenix LiveView.
- Using a mobile device as a full node when the mobile device is in my home. If the UI catches up with features like community administration, it would allow eliminating the need for the combo desktop app + mobile app, and some users could appreciate that. When I say mobile device I also include tablets in the equation. Besides, foldable phones might become more of a thing.
- Could be a nice push in the architectural direction of keeping clients thin (nowadays the mobile client for example has too many responsibilities that should be in status-go IMO). With truly thin clients, the community would be able to more easily fork the clients and build their own versions.
- Running status-go separately would give some users the power to more easily debug release builds, using actual data (I mean not just for development).

Of course, this capability would be optional and not the default. If we could make the release build configurable to support the HTTP backend it would be 🔥 Or is it already? The default for release builds would still be to embed as a shared lib, I fully agree with you 👍🏼 

Curious to hear your thoughts :)

cc @Samyoul @osmaczko @qfrank

**Comment by qfrank** [comment]
> > There's no need to use the whole TCP/IP/HTTP stack for this when we can call the method directly. It will be slower with the network and OS involved.
> > I suggest to make this optional, used only in local debug builds.
> 
> Fully agree @igor-sirotin. But, sorry there's always a but :) I wonder (on a very personal level) if running status-go in a separate machine in my LAN would work well.
> 
> The overhead could be compensated by some other advantages. I always dreamed of having Status clients running as simpler, thin shells, which could help with some use cases:
> 
> * status-go could run much faster and wouldn't be restricted by the hardware the client is running on.
> * The overhead of the extra layers could be negligible if latency is low enough (in a LAN for example). Reminds me of Phoenix LiveView.
> * Using a mobile device as a full node when the mobile device is in my home. If the UI catches up with features like community administration, it would allow eliminating the need for the combo desktop app + mobile app, and some users could appreciate that. When I say mobile device I also include tablets in the equation. Besides, foldable phones might become more of a thing.
> * Could be a nice push in the architectural direction of keeping clients thin (nowadays the mobile client for example has too many responsibilities that should be in status-go IMO). With truly thin clients, the community would be able to more easily fork the clients and build their own versions.
> * Running status-go separately would give some users the power to more easily debug release builds, using actual data (I mean not just for development).
> 
> Of course, this capability would be optional and not the default. If we could make the release build configurable to support the HTTP backend it would be 🔥 Or is it already? The default for release builds would still be to embed as a shared lib, I fully agree with you 👍🏼
> 
> Curious to hear your thoughts :)
> 
> cc @Samyoul @osmaczko @qfrank

HAHA, We have the same dream, let's start with local debug build first

File: mobile/status.go
Comment (qfrank) [reply mobile/status.go:288]: `callWithResponse` will log password, `callWithResponse` will only take care of json params, `VerifyAccountPasswordV2` will replace `VerifyAccountPassword`

File: protocol/requests/change_database_password.go
Comment (qfrank) [reply protocol/requests/change_database_password.go]: fixed

File: protocol/requests/connection_change.go
Comment (qfrank) [reply protocol/requests/connection_change.go]: fixed

File: protocol/requests/convert_to_keycard_account.go
Comment (qfrank) [reply protocol/requests/convert_to_keycard_account.go]: fixed

File: protocol/requests/input_connection_string_for_bootstrapping.go
Comment (qfrank) [reply protocol/requests/input_connection_string_for_bootstrapping.go]: except we want touch the old logic which is a bit complicated🤔 I would prefer keep it as string here to keep it stable

Comment (igor-sirotin) [reply protocol/requests/input_connection_string_for_bootstrapping.go]: I think it's better to fix right away. Otherwise it will stay like this forever. We would have to implement a V3 endpoint to change it 😄

Comment (igor-sirotin) [reply protocol/requests/input_connection_string_for_bootstrapping.go]: Maybe merge all other endpoints first, and work with this one as a separate PR?

Comment (qfrank) [reply protocol/requests/input_connection_string_for_bootstrapping.go]: let me update it in this PR if you have a strong opinion :)

**Comment by ulisesmac** [comment]
> To get started, there's a README in `/cmd/status-backend/README.md`.
> 
> ```shell
> make generate
> make run-status-backend PORT=9050
> ```

I'm new to status-go, so I need some help,
when I run `make run-status-backend ...`
I always get the following output at the end:

```
go run ./cmd/status-backend --address localhost:12345
Configuring Nix shell for target 'default'...
vendor/github.com/waku-org/go-zerokit-rln/rln/link/x86_64.go:8:8: cannot find module providing package github.com/waku-org/go-zerokit-rln-x86_64/rln: import lookup disabled by -mod=vendor
	(Go version in go.mod is at least 1.14 and vendor directory exists.)
make: *** [Makefile:181: run-status-backend] Error 1
```

Before I built status-go as specified here:
https://github.com/status-im/status-go/blob/develop/_docs/how-to-build.md#build-status-go

What is the step I'm missing out? :thinking:

**Comment by qfrank** [comment]
> > To get started, there's a README in `/cmd/status-backend/README.md`.
> > ```shell
> > make generate
> > make run-status-backend PORT=9050
> > ```
> 
> I'm new to status-go, so I need some help, when I run `make run-status-backend ...` I always get the following output at the end:
> 
> ```
> go run ./cmd/status-backend --address localhost:12345
> Configuring Nix shell for target 'default'...
> vendor/github.com/waku-org/go-zerokit-rln/rln/link/x86_64.go:8:8: cannot find module providing package github.com/waku-org/go-zerokit-rln-x86_64/rln: import lookup disabled by -mod=vendor
> 	(Go version in go.mod is at least 1.14 and vendor directory exists.)
> make: *** [Makefile:181: run-status-backend] Error 1
> ```
> 
> Before I built status-go as specified here: https://github.com/status-im/status-go/blob/develop/_docs/how-to-build.md#build-status-go
> 
> What is the step I'm missing out? 🤔

what if install go 1.22 manually?  feel free to DM me on discord @ulisesmac

**Comment by igor-sirotin** [comment]
> ```
> vendor/github.com/waku-org/go-zerokit-rln/rln/link/x86_64.go:8:8: cannot find module providing package github.com/waku-org/go-zerokit-rln-x86_64/rln: import lookup disabled by -mod=vendor
> ```

@ulisesmac you need to build it with `--tags gowaku_no_rln`.

Not sure why it works for everyone else, but we should add it to the command:

```diff
--- a/Makefile
+++ b/Makefile
@@ -178,7 +178,7 @@ status-backend: build/bin/status-backend
 run-status-backend: PORT ?= 0
 run-status-backend: generate
 run-status-backend: ##@run Start status-backend server listening to localhost:PORT
-       go run ./cmd/status-backend --address localhost:${PORT}
+       go run --tags '$(BUILD_TAGS)' ./cmd/status-backend --address localhost:${PORT}
 
 statusd-prune-docker-image: SHELL := /bin/sh
 statusd-prune-docker-image: ##@statusd-prune Build statusd-prune docker image
```

**Comment by osmaczko** [comment]
> * Using a mobile device as a full node when the mobile device is in my home. If the UI catches up with features like community administration, it would allow eliminating the need for the combo desktop app + mobile app, and some users could appreciate that. When I say mobile device I also include tablets in the equation. Besides, foldable phones might become more of a thing.

Hey @ilmotta, I like the idea in general, but I see a few potential challenges and would need further clarification on your vision. For instance, what happens when you're away from home? Would the mobile app still function in that case? If so, how would it operate without a connection to the status-go backend?

If the connection to the status backend is designed to intelligently switch to a LAN version when at home, there are a few concerns. First, switching from a restricted mobile version to a full-node experience could introduce significant complexity on both the backend and UI side. Second, typical home wifi routers can become unstable if overwhelmed by the high volume of traffic generated by running a full node. I'm not referring to the messages propagated by Waku itself, but rather the logic within status-go. For example, in an abstract scenario like importing a community from an archive, thousands of signals (websocket messages in this case) could be sent per millisecond, potentially overwhelming the router. That said, the latter is just a guess—perhaps this wouldn't be an issue in practice.

**Comment by ilmotta** [comment]
Hey @osmaczko, you are raising great questions. Thanks for coming back to this subject.

I haven't thought too much about technical feasibility or even about how the experience would be. I guess the use case I described can be useful, but only to a tiny 🪡 fraction of the Status user base. For everything I say here, please take it lightly, I'm not representing anything from Status :)

There are a few constraints from my particular context:

- I work from home the majority of the time.
- I have access to fiber and I believe bandwidth wouldn't be a problem for my plan.
- I'd be willing to upgrade my router if I'm limited by it.
- I'd be willing to pay for running the node in the cloud. Maybe my wallet won't like it 😅 I'd be of course very concerned about security, so I wouldn't use this for important stuff until "proven safe".
- I have a few machines to spare at home. Running a full node wouldn't affect my work machines (except maybe overloading the network, but there are ways around this).

I mentioned about the local network use case, but the most interesting one to me would be to run the server in the cloud or publicly exposed via my own local infra. If we eventually solve the problem of good incentives for running full nodes, the infra cost could be partially offset.

In order to help justify the operational cost, as a user, I would like to have access to all possible Status features on "any" device. But this is not possible nowadays, clients have rightfully being developed with platform restrictions in mind, thus the mobile client is restricted. It's kind of a chicken and egg problem.

One other scenario I thought before was about having one status-go for multiple clients. Let's say, when running the desktop app at home, that "instance" of status-go could be the main one managing my other clients (in case they are paired and only if they're in the same local network), sort of a creating a seamless experience to sync data between them instantly without relying exclusively on the Waku layer. Okay, this is kind of a stretch, would complicate many things. Just sharing because why not :)

If we keep brainstorming, new use cases will pop-up. Whether they can justify cost of development is a completely different story.

> For instance, what happens when you're away from home? Would the mobile app still function in that case? If so, how would it operate without a connection to the status-go backend?
>
> First, switching from a restricted mobile version to a full-node experience could introduce significant complexity on both the backend and UI side.

The UX aspect I would be totally fine doing something manually. For example, while outdoors, the mobile app would detect the abrupt connection drop and enter a "frozen" state where I would have to manually switch to the standard light mode and restart the app.

A more drastic measure could be to have two completely distinct apps, compiled quite differently even. I would restore the same Status profiles in both apps. One app would be the standard, good for outdoors usage, the other would only work at home and with more features. But there are many problems with this approach.

The experience side of things I think is solvable, especially because it would be a niche part of the product, so users would be willing to make some sacrifices. On the infra side, we could help by giving users quick ways to securely deploy. I see Syncthing as a great example in the FOSS world.

> For example, in an abstract scenario like importing a community from an archive, thousands of signals (websocket messages in this case) could be sent per millisecond, potentially overwhelming the router.

I would expect websockets to handle at least a handful of thousands of messages per second even in cheap desktop machines. The router I can't say much. Probably in the case you described, a mobile client would be overwhelmed way before the router. I lack context here, but a solution could be to apply backpressure correctly in the importing process (limit throughout dynamically). Added complexity as you said 👍🏼 

One thing I'm curious about is to understand how much worse (if worse at all in practice) would the experience be to users, if status-backend was the only way of integrating with status-go.

I sincerely believe the complexity in the backend is very high and with many unknowns. And there's no strong evidence to justify this effort. But devs might gradually explore this architecture on their own. I started to think more about these problems when I attempted to create a messaging app within Emacs, without all the cruft in the desktop/mobile clients. Emacs is single-threaded and it's hard to optimize some things and I wanted to keep the client as light as possible. I'd pay to use Status like that :) Similar to this client, for example https://github.com/alphapapa/ement.el

**Review by igor-sirotin — APPROVED** [review]

**Review by ilmotta — APPROVED** [review]

**Review by igor-sirotin — CHANGES_REQUESTED** [review]
I have some concerns regarding `MEDIA_HTTPS` thingy, please take a look at https://github.com/status-im/status-go/pull/6060#pullrequestreview-2442755942.

Also, I highly recommend to split this PR into fewer smaller ones. I have already approved this PR 3 weeks ago, an you could have merged it back then, but the PR evolved and there're now more changes on top. And now new changes are holding old ones, although they're technically not dependent on each other. And this is not even counting the burden of reviewers to read so many changes.

**Comment by qfrank** [comment]
> I have some concerns regarding `MEDIA_HTTPS` thingy, please take a look at [#6060 (review)](https://github.com/status-im/status-go/pull/6060#pullrequestreview-2442755942).
> 
> Also, I highly recommend to split this PR into fewer smaller ones. I have already approved this PR 3 weeks ago, an you could have merged it back then, but the PR evolved and there're now more changes on top. And now new changes are holding old ones, although they're technically not dependent on each other. And this is not even counting the burden of reviewers to read so many changes.

Let me explain it a bit as you might don't have the full context.
You can review it by commit, you have approved this PR already, I didn't expect you review it again ... You can just review the later [one](https://github.com/status-im/status-go/pull/6060). It doesn't mean I can merge it once I got 2 approvals, I need to pass manual QA. The relate mobile PRs ( [21450](https://github.com/status-im/status-mobile/pull/21450) / [21550](https://github.com/status-im/status-mobile/pull/21550) ) need full regression test, so we had a conclusion to combile these PRs to test once.
Maybe they're technically not dependent on each other in your point view, These PRs all aim to achieve the same goal: ensuring the mobile frontend can properly use the status backend server. Since this PR hasn't been merged yet, I need to test the status backend server locally. To address the HTTPS issue, I must create a separate PR based on this one. From the frontend's perspective, PRs 5943 and 6060 are functionally related. @igor-sirotin

**Comment by igor-sirotin** [comment]
@qfrank I see you intention, but I still think it should've been done a bit differently.
I guess at this point it doesn't make sense to change, but I'll post it for the future PRs.

> ... I need to pass manual QA

I think this is a bit wrong approach in this case.

There're 2 things to be tested:
- New `status-go` code
- Integration to `status-mobile`

Because 99% of changes in this PR is new code, you're not affecting desktop/mobile old behaviour.

So It is the integration to `status-mobile` that requires manual QA and full regression test. But the `status-go` changes can be merged independently with it's own testing.

Maybe you're using `status-mobile` regression tests to test `status-go` changes, but this is also not quite right. `status-go` should be self-tested, rather then with `status-mobile`. I know functional tests framework is still not in production condition, so I was ready to merge it without coverage. Just if you have tested each new endpoint manually (e.g. with Postman) would be ok in this case.

**Comment by qfrank** [comment]
> @qfrank I see you intention, but I still think it should've been done a bit differently. I guess at this point it doesn't make sense to change, but I'll post it for the future PRs.
> 
> > ... I need to pass manual QA
> 
> I think this is a bit wrong approach in this case.
> 
> There're 2 things to be tested:
> 
>     * New `status-go` code
> 
>     * Integration to `status-mobile`
> 
> 
> Because 99% of changes in this PR is new code, you're not affecting desktop/mobile old behaviour.
> 
> So It is the integration to `status-mobile` that requires manual QA and full regression test. But the `status-go` changes can be merged independently with it's own testing.
> 
> Maybe you're using `status-mobile` regression tests to test `status-go` changes, but this is also not quite right. `status-go` should be self-tested, rather then with `status-mobile`. I know functional tests framework is still not in production condition, so I was ready to merge it without coverage. Just if you have tested each new endpoint manually (e.g. with Postman) would be ok in this case.

make sense, pls merge it, thank you @igor-sirotin

**Comment by igor-sirotin** [comment]
Merging now. `MEDIA_HTTPS` will be removed in an upcoming PR, details in https://github.com/status-im/status-go/pull/6060#discussion_r1847978959

**Pull request merged**
